web/src/components/pages/contribution/listen/listen.tsx

```diff
@@ -173,6 +172,16 @@ class ListenPage extends React.Component<Props, State> {
       // Each user can only get once.
       api.setInviteContributeAchievement();
     }
+  };
+
+  private vote = (isValid: boolean) => {
+    const { clips } = this.state;
+    const clipIndex = this.getClipIndex();
+
+    this.stop();
+    Promise.all([
+      this.props.vote(isValid, this.state.clips[this.getClipIndex()].id),
+    ]).then(this.showAchievementToast);
```

Comment:
It looks like there's only one promise in here; is this equivalent to:

```
this.props.vote(isValid, this.state.clips[this.getClipIndex()].id).then(this.showAchievementToast);
```

?

Reply:
Yes, only typescript will complain about strange syntax error. It treat the then() as somewhat a property.

Reply:
I'm looking into this now; I see what you're referring to:

```
Property 'then' does not exist on type '(dispatch: Dispatch<LoadAction | RefillCacheAction | RemoveClipAction | AchievementAction | UpdateAction | TallyRecordingAction | TallyVerificationAction>, getState: () => StateTree) => Promise<...>'.
```

It seems that `this.props.vote` returns an async function, not a Promise. I think the `Promise.all()` might be obscuring an actual error.

Reply:
feel free to change/refactor anything.
I don't know that much about React.

Reply:
From https://react-redux.js.org/5.x/api/connect#docsNav:

> `[mapDispatchToProps(dispatch, [ownProps]): dispatchProps]` (Object or Function): If an object is passed, each function inside it is assumed to be a Redux action creator. An object with the same function names, **but with every action creator wrapped into a dispatch call so they may be invoked directly,** will be merged into the component’s props.

The types we're using for these action creators are:

```typescript
interface PropsFromDispatch {
  removeClip: typeof Clips.actions.remove;
  vote: typeof Clips.actions.vote;
  addAchievement: typeof Notifications.actions.addAchievement;
}
```

They seem to be referencing the _original_ functions, not the wrapped functions that can be invoked directly. That explains the error.

My suggestion would be to rewrite the types pf `PropsFromDispatch` as needed, eg:

```typescript
interface PropsFromDispatch {
  removeClip: (clipId: string) => Promise<any>;
  vote: (isValid: boolean, clipId?: string) => Promise<any>;
  addAchievement: (
    score: number,
    text: string,
    type: NotificationType
  ) => Promise<any>;
}
```

@Gregoor you've probably encountered this problem heaps before. Any guidance?

Reply:
@rileyjshaw , the code can run, I think it's more of a programming style, a guideline thing. To rewrite the types of `PropsFromDispatch `, we need to change  many places, causing nonresistant interface style of different `PropsFromDispatch `.
Maybe get a coding guideline first, stating things like: prefer `Promise `to `async`. So that the contributors could know the direction.

Reply:
I can't seem to recall how we handled async functions in the past. My expectation would've been that redux(-thunk) respects an action's async-ness when `connect`ing it. But I might be expecting something from those libs that's maybe not even possible to express in TS.
Anyway, when types stand in your way I also vote for what Riley said: override the types to make it fit (and reconsider things more fundamentally when it becomes a pattern).

Reply:
OK, but how about the minimum modifications that I just pushed here: https://github.com/mozilla/voice-web/pull/2392/commits/fdf5d551c433ba2f7d3a98e4fe9615a339235ab5?

It is small change, it's working and it doesn't have introduce Promise here.